src/signing_key.py

```diff
@@ -0,0 +1,115 @@
+#!/usr/bin/env python3
+
+# Copyright 2025 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Synapse signing key."""
+
+
+import logging
+import typing
+
+import ops
+
+import synapse
+from charm_state import CharmState
+
+logger = logging.getLogger(__name__)
+
+
+def signing_key_path(charm_state: CharmState) -> str:
+    """Get signing key path.
+
+    Args:
+        charm_state: charm state.
+
+    Returns:
+        signing key path as string.
+    """
+    return f"/data/{charm_state.synapse_config.server_name}.signing.key"
+
+
+def get_signing_key_secret_content(
+    peer_relation: typing.Optional[ops.Relation], charm: ops.CharmBase
+) -> typing.Optional[str]:
+    """Get signing key secret content.
+
+    Args:
+        peer_relation: synapse peer relation.
+        charm: charm instance.
+
+    Returns:
+        Content as string.
+    """
+    content = None
+    if not peer_relation:
+        logger.error(
+            "Failed to get signing key: no peer relation %s found",
+            synapse.SYNAPSE_PEER_RELATION_NAME,
+        )
+        return content
+    secret_id = peer_relation.data[charm.app].get("secret-signing-id")
+    if secret_id:
+        try:
+            secret = charm.model.get_secret(id=secret_id)
+            content = secret.get_content().get("secret-signing-key")
+        except (ops.model.SecretNotFoundError, ValueError, TypeError) as exc:
+            logger.exception("Failed to get secret id %s: %s", secret_id, str(exc))
+            del peer_relation.data[charm.app]["secret-signing-id"]
+    return content
+
+
+def write_to_container(
+    peer_relation: typing.Optional[ops.Relation],
+    charm: ops.CharmBase,
+    charm_state: CharmState,
+    container: ops.Container,
+) -> None:
+    """Get signing key from secret and write to container.
+
+    Args:
+        peer_relation: synapse peer relation.
+        charm: charm instance.
+        charm_state: charm state.
+        container: container.
+    """
+    content = get_signing_key_secret_content(peer_relation, charm)
+    if content:
+        container.push(
+            signing_key_path(charm_state),
+            content,
+            make_dirs=True,
+            encoding="utf-8",
+        )
+
+
+def write_to_secret(
+    peer_relation: typing.Optional[ops.Relation],
+    charm: ops.CharmBase,
+    charm_state: CharmState,
+    container: ops.Container,
+) -> None:
+    """Create secret with signing key content.
+
+    Args:
+        peer_relation: synapse peer relation.
+        charm: charm instance.
+        charm_state: charm state.
+        container: container.
+    """
+    if not peer_relation:
```

Comment:
Either done here and in `write_to_container` or done in `get_signing_key_secret_content` but not necessary for both